api/include/opentelemetry/trace/trace_state.h

```diff
@@ -0,0 +1,107 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+
+// TraceState carries tracing-system specific context in a list of key-value pairs. TraceState
+// allows different vendors to propagate additional information and inter-operate with their legacy
+// Id formats.
+//
+// Implementation is optimized for a small list of key-value pairs.
+//
+// Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+// and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+// forward slashes /.
+//
+// Value is opaque string up to 256 characters, of printable ASCII RFC0020 characters (i.e. the
+// range 0x20 to 0x7E) except comma , and equals =.
+class TraceState
+{
+public:
+  static constexpr int kKeyMaxSize       = 256;
+  static constexpr int kValueMaxSize     = 256;
+  static constexpr int kMaxKeyValuePairs = 32;
+
+  class Entry
+  {
+  public:
+    virtual ~Entry() {}
+
+    virtual nostd::string_view key()   = 0;
+    virtual nostd::string_view value() = 0;
+  };
+
+  // An empty TraceState.
+  TraceState() noexcept = default;
+
+  virtual ~TraceState() = default;
+
+  // Returns false if no such key, otherwise returns true and populates value.
+  virtual bool Get(nostd::string_view key, nostd::string_view *value) const noexcept
+  {
+    return false;
+  }
+
+  // Returns true if there are no keys.
+  virtual bool empty() const noexcept { return true; }
+
+  // Returns a span of all the entries. The TraceState object must outlive the span.
+  virtual nostd::span<Entry *> entries() const noexcept { return {}; }
+
+  // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
+  // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+  // forward slashes /.  For multi-tenant vendor scenarios, an at sign (@) can be used to prefix the
+  // vendor name.
+  static bool IsValidKey(nostd::string_view key)
+  {
+    if (key.empty() || key.length() > kKeyMaxSize || !IsNumberOrDigit(key[0]))
+    {
+      return false;
+    }
+    int ats     = 0;
+    const int n = key.length();
+    for (int i = 0; i < n; ++i)
+    {
+      char c = key[i];
+      if (!IsNumberOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/')
```

Comment:
Vendor name has a size limitation of 13 characters according to the [spec](https://www.w3.org/TR/trace-context/#key).